### PR TITLE
Improve path generators

### PR DESCRIPTION
**Pull request by mossprescott**
The incoming `size` parameter is now respected, and the available size is distributed among the generated path segments. Crucially, that means `/` and `.` can now be generated (and *alway* are, when size = 0), along with simple paths like `/a` and also much deeper paths, all under the control of the caller.

Biased the segment generator to produce mostly simple, alpha-numeric path segments, with arbitrary Unicode characters appearing less often. This just makes it easier to read the resulting paths most of the time, without sacrificing coverage.

### CODE REVIEW

File: scalacheck/src/main/scala/pathy/scalacheck/package.scala [hunk 1]
```diff
@@ -18,9 +18,43 @@ package pathy
...
+    Gen.size.flatMap { n =>
+      if (n < 1) Gen.const(Nil)
+      else if (n < 3) g.map(_ :: Nil)
+      else
+        for {
+          l <- Gen.choose(1, n)
+          r = n - l - 1
+          h <- Gen.resize(l, g)
+          t <- Gen.resize(r, sizeDistributedListOfNonEmpty(g))
+        } yield (h :: t)
```
Comment (mossprescott) [inline scalacheck/src/main/scala/pathy/scalacheck/package.scala]: @jdesah: Poster child for Computation Expressions?

Abandoning the `for` comprehension helps a bit, but it's not easy for me to follow that way:
```scala
Gen.choose(1, n).flatMap(l => 
  (Gen.resize(l, g) |@| 
    Gen.resize(n - l - 1, sizeDistributedListOfNonEmpty(g)))(_ ::_)
```

Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/package.scala]: :+1: And it's a pretty easy looking one, I don't see any reason it wouldn't work with the current version.

**Comment by mossprescott** [comment]
@jdesah: this was motivated when I found that `/` never seemed to be generated and I looked into why. 

I think `Gen.resize(<a constant>)` is something we should *never* do, and it was done here of course to work around the problem of generating lists of *n* things which are all *n*-sized, which blows up quickly. I set out to fix that in a sensible way, at least in this setting, and `sizeDistributed...` is the result. With these generators, our tests will start with nice simple paths and work up to paths containing multiple segments and up to 100 (or is it 200) *total characters/delimiters*.

File: scalacheck/src/main/scala/pathy/scalacheck/RandomSeg.scala [hunk 2]
```diff
@@ -28,9 +28,10 @@ private[scalacheck] object RandomSeg {
   implicit val randomSegArbitrary: Arbitrary[RandomSeg] =
     Arbitrary {
       Gen.nonEmptyListOf(Gen.frequency(
-        100 -> Arbitrary.arbitrary[Char],
+        100 -> Gen.alphaNumChar,
```
Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/RandomSeg.scala]: This is the reason @wemrysi and I introduced the `PathOf` construct so that you can say `PathOf[AlphaCharacters]`, but I think the default should remain a truly arbitrary path. Also, I defined `AlphaCharacters` in quasar, might make sense to pull it into pathy as it seems like a generally useful thing.

File: scalacheck/src/main/scala/pathy/scalacheck/RandomSeg.scala [hunk 3]
```diff
@@ -28,9 +28,10 @@ private[scalacheck] object RandomSeg {
...
          10 -> Gen.const('.'),
-         10 -> Gen.const('/')
+         10 -> Gen.const('/'),
+          5 -> Arbitrary.arbitrary[Char]
```
Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/RandomSeg.scala]: Ah, I spoke too soon I guess. This line does mean that we preserve the semantic of a random segment. I could be convinced this change in relative frequencies is for the better.

File: scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala [hunk 4]
```diff
@@ -54,19 +54,14 @@ object PathOf {
     genRelFile[A] map (rootDir </> _)
 
   private def genRelFile[A: Arbitrary: Show]: Gen[RelFile[Sandboxed]] =
-    for {
-      d <- genRelDir[A]
-      s <- genSegment[A]
-    } yield d </> file(s)
+    sizeDistributed(genRelDir[A], genSegment[A])((d, s) => d </> file(s))
```
Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala]: So here if the `size` parameter is set to 10, in the old version, we would get a path with 10 directories with 10 characters each and a filename of 10 characters?

File: scalacheck/src/main/scala/pathy/scalacheck/package.scala [hunk 5]
```diff
@@ -18,9 +18,43 @@ package pathy
...
+  /** Generator for lists of non-atomic components, where the size parameter is
+    * spread across all the generated elements so that the aggregate size of
+    * component/leaf elements is effectively controlled by the size parameter.
+    * No element is ever generated with a size parameter of less than 1, and 
+    * each "cons cell" consumes one unit of size.
+    */
+  def sizeDistributedListOfNonEmpty[A](g: Gen[A]): Gen[List[A]] =
+    Gen.size.flatMap { n =>
+      if (n < 1) Gen.const(Nil)
+      else if (n < 3) g.map(_ :: Nil)
```
Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/package.scala]: `n < 3`? I was expecting `n == 1`.

File: scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala [hunk 6]
```diff
@@ -54,19 +54,14 @@ object PathOf {
...
 
   private def genAbsDir[A: Arbitrary: Show]: Gen[AbsDir[Sandboxed]] =
     genRelDir[A] map (rootDir </> _)
 
   private def genRelDir[A: Arbitrary: Show]: Gen[RelDir[Sandboxed]] =
```
Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala]: `size` is a very new concept to me in scalacheck, but currently I believe that this generator as implemented will not return a path with 10 segments if the `size` parameter is set to 10. If the behavior for `Gen[List[A]]` is to generate a list of 10 elements if the `size` parameter is set to 10, then I would say the current behavior is not optimal.

File: scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala (hunk 4 above)
Comment (mossprescott) [inline scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala]: Yes. If you just use `nonEmptyListOf` as we did before, you get lists of up to `size` segments, each with up to `size` chars. The problem is that by default scalacheck/specs2 start with small sizes and then go as high as 100/200, which means paths of thousands of chars. On the other hand, you don't want to override the size to something hard-coded because it means you don't get the nice behavior of failing on nice small values first.

The idea of this "aggregate" size stuff is for `size` to control the overall size. For example, with 0, this always generates `/` (for absDir), with 1, you get `/a/`, 2 -> `/ab/`, 3 -> `/abc/` or `/a/b/`, etc. And the paths grow linearly with the size param, which means no "reasonable" size will generate an unreasonably large path.

File: scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala (hunk 6 above)
Comment (mossprescott) [inline scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala]: Not sure what you mean by optimal. With a size of 10, the generator can produce a path with as many as 5 segments, where each dir/filename is a single char, since the `/`s are effectively counted.

File: scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala (hunk 4 above)
Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala]: Ah, I understand better now.

File: scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala (hunk 6 above)
Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/PathOf.scala]: Fair enough. Ya that makes more sense now with your previous explanation.

File: scalacheck/src/main/scala/pathy/scalacheck/package.scala (hunk 5 above)
Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/package.scala]: This makes sense now with your previous explanation.

Comment (mossprescott) [inline scalacheck/src/main/scala/pathy/scalacheck/package.scala]: If n = 2 then head will consume at least 1 and 1 is reserved for `::`, so there will never be enough size to generate a second element and what this really accomplishes is forcing a size-2 head element to be generated.

File: scalacheck/src/main/scala/pathy/scalacheck/RandomSeg.scala (hunk 2 above)
Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/RandomSeg.scala]: @mossprescott You answered my numerous questions and I think this PR is ready to be merged. There is still something that itches me though with this particular change. I wonder if we could create a `Arbitrary[PrettySeg]` or maybe just stay like this but bring up a bit the relative frequency of the truly arbitrary `Char`. I am concerned we are likely to miss some corner cases with such a generator.

Also somewhat unrelated thought, I wonder if it would be nicer if we had `20 -> Gen.chooseOneOf(escapedChars)` where `escapedChars = List('.', '/')`?

Comment (mossprescott) [inline scalacheck/src/main/scala/pathy/scalacheck/RandomSeg.scala]: What corner cases do you have in mind?

What I actually wanted was something like `100 -> Gen.asciiChar`, but there isn't any such thing as far as I could tell, so I went with `alphaNum` as the best available substitute. If you like that better, I could try to come up with a succinct way to express it. I guess I would still want to filter out things like `\n` and `\t`, though, just because they're hard to read in the test output.

I agree giving a name to the list of escaped chars is helpful.

Comment (jedesah) [inline scalacheck/src/main/scala/pathy/scalacheck/RandomSeg.scala]: All I know is that when converting web into the Free architecture, we also transformed most of the tests into properties and that exposed many bugs that were latent in the previous implementation. So much so, that I had to go with `PathOf[AlphaCharacters]` in many places just to move on and complete the translation in a timely manner. I don't know if we will ever find the time to change the properties back to `PathOf[RandomSeg]` or whether that makes sense, but my concern is that we might not do as good of a job exposing problems as many of the tests would only break every so often, so presumably the corner cases were hard to find/generate. On the other hand, I do believe they were probably mostly caused by the `escapedChars` so I can't say for sure that keeping many truly arbitrary characters would help find these corner cases.

I don't have any strong argument against this change and the important thing for me is that it's relatively easy to customize so I don't want to spend an exaggerated amount of time discussing what is the most appropriate default behavior. So as far as I am concerned, I'm okay with this change as is.
